src/Bicep.LangServer/Registry/ModuleRestoreScheduler.cs

```diff
@@ -0,0 +1,134 @@
+        /// Does not wait for the operation to complete and returns immediately.
+        /// </summary>
+        /// <param name="references">The module references</param>
+        public void RequestModuleRestore(ICompilationManager compilationManager, DocumentUri documentUri, IEnumerable<ModuleDeclarationSyntax> references)
+        {
+            this.CheckDisposed();
+            var item = new QueueItem(compilationManager, documentUri, references.ToImmutableArray());
+            lock (this.queue)
+            {
+                this.queue.Enqueue(item);
+
+                // notify consumer about new items
+                this.manualResetEvent.Set();
+            }
+        }
+
+        public void Start()
+        {
+            this.CheckDisposed();
+            this.consumerTask = Task.Factory.StartNew(this.ProcessQueueItems, this.cancellationTokenSource.Token, TaskCreationOptions.LongRunning, TaskScheduler.Default);
+        }
+
+        public void Dispose()
+        {
+            // this is a sealed class - no need for full IDisposable implementation
+            if(!this.disposed)
+            {
+                this.disposed = true;
+                if (this.consumerTask is not null)
+                {
+                    this.cancellationTokenSource.Cancel();
+                    this.consumerTask = null;
+                }
+            }
+        }
+
+        private void ProcessQueueItems()
+        {
+            var token = this.cancellationTokenSource.Token;
+            while (true)
+            {
+                this.manualResetEvent.Wait(token);
+
+                var notifications = new HashSet<CompletionNotification>();
+                var references = new List<ModuleDeclarationSyntax>();
+                lock (this.queue)
+                {
+                    this.UnsafeCollectModuleReferences(notifications, references);
+                    Debug.Assert(this.queue.Count == 0, "this.queue.Count == 0");
+
+                    // queue has been consumed - next iteration should block until more items have been added
+                    this.manualResetEvent.Reset();
+                }
+
+                // this blocks until restore is completed
+                // the dispatcher stores the results internally and manages their lifecycle
+                token.ThrowIfCancellationRequested();
+                if(!this.moduleDispatcher.RestoreModules(references))
+                {
+                    // nothing needed to be restored
+                    // no need to notify about completion
+                    continue;
+                }
+
+                // notify compilation manager that restore is completed
+                // to recompile the affected modules
+                token.ThrowIfCancellationRequested();
+                foreach (var notification in notifications)
+                {
+                    notification.CompilationManager.RefreshCompilation(notification.Uri);
```

Comment:
Storing a list of URIs without module context and notifying the compilation manager on refresh feels like it could be prone to race conditions, and unnecessary refreshes. Feels like we could simplify the design by instead notifying the compilation manager of the list of the modules that have been restored, and have it figure out the files to refresh at that point in time - in a thread-safe manner, and with smarts to figure out the correct order to recompile syntax trees to minimize cascading compilations.

Reply:
Yeah I like this idea. Moving it over there makes it possible for the compilation manager to refresh based on its latest state.

Reply:
I started looking at it, but I may end up changing the contract more than I thought originally. The order part can't really be done until we start using semantic models (I know you started on the fix.) and I've got code split between two branches right now which will make the change hard to test. Do you mind if I do this in a separate PR?

Reply:
Without a remote registry implementation registered, this code won't really be triggered anyway, so no race condition... yet.